asllib/doc/asl.spec

```diff
@@ -1295,10 +1296,3128 @@ typing function annotate_literal(tenv: static_envs, l: literal) -> (t: ty)
+  yielding a list of constraints --- {annotated_cs}.
+  If the list is empty, the result is either
+  $\CannotUnderapproximate$ or $\CannotOverapproximate$,
+  based on {approx} (this function is invoked in the
+  context of approximating lists of constraints).
+  \ProseOtherwiseTypeError",
+  prose_application = ""
+};
+
+relation binop_filter_rhs(
+    approx: constants_set(Over,Under),
+    tenv: static_envs,
+    op: binop,
+    cs: list0(int_constraint)) ->
+         (new_cs: list1(int_constraint)) | constants_set(CannotUnderapproximate,CannotOverapproximate)
+{
+  "filters the list of constraints {cs} by removing
+  values that will definitely result in a dynamic error
+  if found on the right-hand-side of a binary operation
+  expression with the operator {op} in any environment
+  consisting of the \staticenvironmentterm{} {tenv}. The
+  result is the filtered list of constraints {new_cs}.
+  If the list is empty, the result is either
+  $\CannotUnderapproximate$ or $\CannotOverapproximate$,
+  based on {approx} (this function is invoked in the
+  context of approximating lists of constraints).",
+  prose_application = ""
+};
+
+relation refine_constraint_by_sign(tenv: static_envs, p: fun Z -> Bool, c: int_constraint) ->
+         (c_opt: option(int_constraint))
+{
+  "takes a predicate {p} that returns $\True$ based on
+  the sign of its input. The function conservatively
+  refines the constraint {c} in {tenv} by applying
+  symbolic reasoning to yield a new constraint (inside
+  an optional) that represents the values that satisfy
+  the {c} and for which {p} holds. In this context,
+  conservatively means that the new constraint may
+  represent a superset of the values that a more precise
+  reasoning may yield. If the set of those values is
+  empty the result is $\None$.",
+  prose_application = ""
+};
+
+relation reduce_to_z_opt(tenv: static_envs, e: expr) ->
+         (z_opt: option(Z))
+{
+  "returns an integer inside an optional if {e} can be
+  symbolically simplified into an integer in {tenv} and
+  $\None$ otherwise. The expression {e} is assumed not
+  to yield a \typingerrorterm{} when applying
+  $\normalize$ to it.",
+  prose_application = ""
+};
+
+relation refine_constraints(
+    approx: constants_set(Over,Under),
+    tenv: static_envs,
+    f: fun int_constraint -> option(int_constraint),
+    cs: list0(int_constraint)) ->
+         (new_cs: list0(int_constraint)) | constants_set(CannotUnderapproximate,CannotOverapproximate)
+{
+  "refines a list of constraints {cs} by applying the
+  refinement function {f} to each constraint and
+  retaining the constraints that do not refine to
+  $\None$. The resulting list of constraints is given in
+  {new_cs}. If the list is empty, the result is either
+  $\CannotUnderapproximate$ or $\CannotOverapproximate$,
+  based on {approx} (this function is invoked in the
+  context of approximating lists of constraints).",
+  prose_application = ""
+};
```

Comment:
There are extra `@` in here
<img width="568" height="232" alt="Screenshot 2025-10-30 at 09 42 10" src="https://github.com/user-attachments/assets/d4ca4e3a-a8ee-43d2-8a68-71b247b5ab92" />